Build failure: freemind

### Nixpkgs version

- Unstable (26.05)

### Steps to reproduce

Add ``freemind`` to ``systemPackages``.

### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/build/323007529

### Relevant log output

```console
freemind> Running phase: buildPhase
freemind> Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF8
freemind> Buildfile: /build/freemind/build.xml
freemind>
freemind> init:
freemind>      [echo] FreeMind Version = 1.0.1, build 6.
freemind>      [echo] Checking for duplicate Resource strings
freemind>
freemind> xmlbind.checkStatusOfGeneration:
freemind>
freemind> gen:
freemind>
freemind> build:
freemind>     [mkdir] Created dir: /build/bin/classes
freemind>
freemind> BUILD FAILED
freemind> /build/freemind/build.xml:157: Unable to find a javac compiler;
freemind> com.sun.tools.javac.Main is not on the classpath.
freemind> Perhaps JAVA_HOME does not point to the JDK.
freemind> It is currently set to "/nix/store/lakdsf2fqlp84ygjq1p5a00avi3kymp7-openjdk-8u472-b08-jre/lib/openjdk/jre"
freemind>
freemind> Total time: 0 seconds
```

### Additional context

_No response_

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.18.12, NixOS, 26.05 (Yarara), 26.05.20260223.2fc6539`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.31.3`
 - channels(root): `"nixos, nixos-hardware"`
 - nixpkgs: `/nix/store/zkrz054gvl1dcxkw25d8lgx8l10px665-source`

### Notify maintainers

@TomaSajt 

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+label%3A%220.kind%3A+build+failure%22).
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/an/ant/package.nix
+++ b/pkgs/by-name/an/ant/package.nix
@@ -2,7 +2,7 @@
   fetchurl,
   lib,
   stdenv,
-  jre,
+  coreutils,
   makeWrapper,
   gitUpdater,
 }:
@@ -10,8 +10,6 @@
 stdenv.mkDerivation (finalAttrs: {
   pname = "ant";
   version = "1.10.15";
-
-  buildInputs = [ jre ];
 
   nativeBuildInputs = [ makeWrapper ];
 
@@ -21,23 +19,60 @@ stdenv.mkDerivation (finalAttrs: {
   };
 
   installPhase = ''
-    mkdir -p $out/share/ant
+    mkdir -p $out/bin $out/share/ant
     mv * $out/share/ant/
 
     # Get rid of the manual (35 MiB).  Maybe we should put this in a
-    # separate output.
-    rm -rf $out/share/ant/{manual,WHATSNEW}
+    # separate output.  Keep the antRun script since it's vanilla sh
+    # and needed for the <exec/> task (but since we set ANT_HOME to
+    # a weird value, we have to move antRun to a weird location).
+    # Get rid of the other Ant scripts since we provide our own.
+    mv $out/share/ant/bin/antRun $out/bin/
+    rm -rf $out/share/ant/{manual,bin,WHATSNEW}
+    mkdir $out/share/ant/bin
+    mv $out/bin/antRun $out/share/ant/bin/
 
-    # Link Ant's special $ANT_HOME/bin to the standard /bin location
-    ln -s $out/share/ant/bin $out
+    cat >> $out/bin/ant <<EOF
+    #! ${stdenv.shell} -e
 
-    # Wrap the wrappers.
-    for wrapper in ant runant.py runant.pl; do
-      wrapProgram "$out/share/ant/bin/$wrapper" \
-        --set-default JAVA_HOME "${jre.home}" \
-        --set-default ANT_HOME "$out/share/ant" \
-        --prefix CLASSPATH : "$out/share/ant/lib"
-    done
+    ANT_HOME=$out/share/ant
+
+    # Find the JDK by looking for javac.  As a fall-back, find the
+    # JRE by looking for java.  The latter allows just the JRE to be
+    # used with (say) ECJ as the compiler.  Finally, allow the GNU
+    # JVM.
+    if [ -z "\''${JAVA_HOME-}" ]; then
+        for i in javac java gij; do
+            if p="\$(type -p \$i)"; then
+                export JAVA_HOME="\$(${coreutils}/bin/dirname \$(${coreutils}/bin/dirname \$(${coreutils}/bin/readlink -f \$p)))"
+                break
+            fi
+        done
+        if [ -z "\''${JAVA_HOME-}" ]; then
+            echo "\$0: cannot find the JDK or JRE" >&2
+            exit 1
+        fi
+    fi
+
+    if [ -z \$NIX_JVM ]; then
+        if [ -e \$JAVA_HOME/bin/java ]; then
+            NIX_JVM=\$JAVA_HOME/bin/java
+        elif [ -e \$JAVA_HOME/bin/gij ]; then
+            NIX_JVM=\$JAVA_HOME/bin/gij
+        else
+            NIX_JVM=java
+        fi
+    fi
+
+    LOCALCLASSPATH="\$ANT_HOME/lib/ant-launcher.jar\''${LOCALCLASSPATH:+:}\$LOCALCLASSPATH"
+
+    exec \$NIX_JVM \$NIX_ANT_OPTS \$ANT_OPTS -classpath "\$LOCALCLASSPATH" \
+        -Dant.home=\$ANT_HOME -Dant.library.dir="\$ANT_LIB" \
+        org.apache.tools.ant.launch.Launcher \$NIX_ANT_ARGS \$ANT_ARGS \
+        -cp "\$CLASSPATH" "\$@"
+    EOF
+
+    chmod +x $out/bin/ant
   '';
 
   passthru = {

